### PR TITLE
Add: highlight Numba IR dump

### LINKED ISSUE #4657: Implement a pygments lexer for Numba IR
## Feature request

It'd be nice, especially in jupyter notebooks/ipython if Numba IR was syntax highlighted. I think most of the work to do this will involve implementing an appropriate lexer for pygments, docs: http://pygments.org/docs/lexerdevelopment/

### PR DESCRIPTION
**Pull request by 1844144**
Closes #4657

Colors in ipython(white background) and Jupyter are ok. `Terminal256Formatter` doesn't recognize black background currently so colors a bit dark on black background.
Demo: 
image

### CODE REVIEW

**Comment by stuartarchibald** [comment]
@1844144 Thanks for the PR. This looks great already! Am going to try it out...!

File: numba/core/ir_lexer.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
```
Comment (eric-wieser) [inline numba/core/ir_lexer.py]: Is this necessary in python 3?

**Comment by stuartarchibald** [comment]
Just tried this out locally, looks good :)

> Closes #4657
> 
> Colors in ipython(white background) and Jupyter are ok. `Terminal256Formatter` doesn't recognize black background currently so colors a bit dark on black background.

Numba has a `color_scheme` config option:
https://github.com/numba/numba/blob/a4c4a8dceb0801672d6fda9199c36047dba16f0b/numba/core/config.py#L143
which guides the colour scheme in its error message reporting. Perhaps this could be used to influence the formatter?

**Comment by stuartarchibald** [comment]
Reference for colour schemes: http://numba.pydata.org/numba-doc/latest/reference/envvars.html#envvar-NUMBA_COLOR_SCHEME

File: numba/core/ir_lexer.py (hunk 1 above)
Comment (1844144) [reply numba/core/ir_lexer.py]: In case our lexer will be included in Pygments  official repo.

Comment (1844144) [reply numba/core/ir_lexer.py]: But you are right, not necessary

**Comment by 1844144** [comment]
Not sure we need to use `colorama`, because `pygments` already has enough styles for different backgrounds (so we can use/tweak it).
https://github.com/pygments/pygments/blob/827e890fe10b4bb8434ffb6350e592cece4b033c/pygments/styles/__init__.py#L17

If it is ok to not to use colorama, I will suggest a style for each case (or even we may allow to select `pygments`'s styles by name along with our predefined styles)

**Comment by stuartarchibald** [comment]
> Not sure we need to use `colorama`, because `pygments` already has enough styles for different backgrounds (so we can use/tweak it).
> https://github.com/pygments/pygments/blob/827e890fe10b4bb8434ffb6350e592cece4b033c/pygments/styles/__init__.py#L17
>
> If it is ok to not to use colorama, I will suggest a style for each case (or even we may allow to select `pygments`'s styles by name along with our predefined styles)

Yes, I was thinking that the styles chosen could just follow Numba's already used names based on output type:
```
    no_color - No color added, just bold font weighting.
    dark_bg - Suitable for terminals with a dark background.
    light_bg - Suitable for terminals with a light background.
    blue_bg - Suitable for terminals with a blue background.
    jupyter_nb - Suitable for use in Jupyter Notebooks.
```
Suggesting some styles would be good thanks.

**Comment by 1844144** [comment]
Suggestions: 

blue_bg -> NativeStyle (not sure my blue background is that what you mean)
image

dark_bg -> MonokaiStyle
image

light_bg -> ManniStyle OR DefaultStyle
image

jupyter_nb -> DefaultStyle
image

no_color -> customized BlackWhiteStyle
image

**Comment by 1844144** [comment]
Few notes:
 - if developer will turn on DUMP_IR he will see black/white style as default value for COLOR_SCHEME is no_color (as for me better would be to have jupyter_nb as default)
 - there are other places, where dumps are highlighted independently from COLOR_SCHEME
https://github.com/numba/numba/blob/90654c5ac525ef7eeaa69f605c15710bc40476de/numba/core/lowering.py#L195-L197
and 
https://github.com/numba/numba/blob/90654c5ac525ef7eeaa69f605c15710bc40476de/numba/core/codegen.py#L36-L39
And code assume you will have **light** background

**Comment by stuartarchibald** [comment]
Thanks for the above. These colour schemes are a good fit for the terminal/output formats we encoded for the error messages. I use dark background terminals, the `MonokaiStyle` looks good and the highlighting is really helpful!

> Few notes:
> 
>     * if developer will turn on DUMP_IR he will see black/white style as default value for COLOR_SCHEME is no_color (as for me better would be to have jupyter_nb as default)

hmmm, is there a theme that renders reasonably on both a dark and a light background? This is what put us off doing anything with colour in the error message highlighting, it wasn't clear how to reasonably accommodate both, so font weighting was applied instead.
 
>     * there are other places, where dumps are highlighted independently from COLOR_SCHEME
>       https://github.com/numba/numba/blob/90654c5ac525ef7eeaa69f605c15710bc40476de/numba/core/lowering.py#L195-L197
>       
>       and
>       https://github.com/numba/numba/blob/90654c5ac525ef7eeaa69f605c15710bc40476de/numba/core/codegen.py#L36-L39
>       
>       There are supposed that you will have **light** background

The above were thrown in during the last release cycle by me as I was working on really involved IR transforms and just wanted something more readable. If it's easy/you are motivated to patch in the `COLOR_SCHEME` stuff into these locations too it would be welcomed, else I think I understand enough about what's going on now to be able to do it :)

Thanks again, this is incredibly useful!!!

**Comment by 1844144** [comment]
Ok, then

1. Modified default theme used as default(no_color) scheme 
image
image

2. Please decide what style should be used for `light_bg`: ManniStyle (as now) or same style as `jupyter_nb` (see images above)

3. Color_scheme in other locations patched

**Comment by stuartarchibald** [comment]
> Ok, then
> 
>     1. Customized DefaultStyle used as default(no_color) scheme

Thanks, this looks good.
 
>     2. Please decide what style should be used for `light_bg`: ManniStyle (as now) or same style as `jupyter_nb` (see images above)

I quite like `ManniStyle`, I guess it's trivial to change later if there are any objections.

>     3. Color_scheme in other locations patched

Brilliant, thanks.

File: numba/misc/dump_style.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            ), suffix=' '), Keyword),
+        ],
+
+        'whitespace': [
+            (r'(\n|\s)', Text),
+        ],
+    }
+
+
+def by_colorscheme():
```
Comment (stuartarchibald) [inline numba/misc/dump_style.py]: Please could you add a docstring, thanks.

File: numba/misc/dump_style.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from pygments.styles.default import DefaultStyle
+from pygments.styles.manni import ManniStyle
+from pygments.styles.monokai import MonokaiStyle
+from pygments.styles.native import NativeStyle
+
+from pygments.lexer import RegexLexer, include, bygroups, words
+from pygments.token import Text, Name, String,  Punctuation, Keyword, \
+    Operator, Number
+
+from pygments.style import Style
```
Comment (stuartarchibald) [inline numba/misc/dump_style.py]: I think it'd be good to have a `try..except ImportError` around this to catch the case where Numba extension writers might try and import this module for use in their code but not have `pygments` installed (as it's only an optional requirement for Numba).  Thinking something like:
https://github.com/numba/numba/pull/5333/files#diff-5de309739e66ef02bd05362a66f0caa1R191-R201

File: numba/core/ir.py [hunk 4]
```diff
@@ -1462,9 +1462,11 @@ def dump(self, file=None):
                     raise ValueError(msg)
                 else:
                     from pygments import highlight
-                    from pygments.lexers import DelphiLexer as lexer
+                    from numba.misc.dump_style \
+                        import NumbaIRLexer as lexer, by_colorscheme
```
Comment (stuartarchibald) [inline numba/core/ir.py]: Perhaps split this as:
```python
from numba.misc.dump_style import NumbaIRLexer as lexer
from numba.misc.dump_style import by_colorscheme
```
the code base layout was recently refactored and it was found that refactoring tools seemed to struggle with more complex spellings like in the above.

File: numba/misc/dump_style.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+
```
Comment (stuartarchibald) [inline numba/misc/dump_style.py]: Empty line? delete?

File: numba/misc/dump_style.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+from pygments.lexer import RegexLexer, include, bygroups, words
+from pygments.token import Text, Name, String,  Punctuation, Keyword, \
+    Operator, Number
+
+from pygments.style import Style
+
+
+class NumbaIRLexer(RegexLexer):
+    """
+    For Numba IR code.
```
Comment (stuartarchibald) [inline numba/misc/dump_style.py]:
```suggestion
    Pygments style lexer for Numba IR (for use with highlighting etc).
```

**Comment by 1844144** [comment]
Done

**Comment by stuartarchibald** [comment]
Thanks for the fixes. I'm happy to merge this as-is as it's a massive improvement on what's in the code base at present! Congratulations on your first contribution to Numba!

In manual testing I did notice that there are some cases which "break", for example:
```python
from numba import config
config.DEBUG=1
from numba import njit
import numpy as np

@njit
def foo(x):
    y = np.sin(x)
    return x + y[0]

foo(np.ones(5))
```
produces a load of output, including an IR dump for `_broadcast_onto`, which highlights incorrectly:

highlighting_issue1
perhaps something to fix in a follow up PR? I also wonder if we can make the Numba IR output easier to parse?!

I also noticed that the LLVM Lexer doesn't "understand" our slightly unusual LLVM IR all the time either, but that's an issue for another time!

**Comment by stuartarchibald** [comment]
@gmarkall wonder if a PTX (and maybe later SASS) lexer would be useful for CUDA dumps?

**Comment by stuartarchibald** [comment]
Note to self. This PR would be good to highlight in the release notebook for 0.49 as it's a great improvement over the existing output and something that should be demonstrated and publicly advertised, it's also obviously extremely helpful for users working with Numba internals.

**Comment by 1844144** [comment]
Thanks for testing, would be good to have most complicated function to test lexer on.

If you will explain your changes(or put a link to the source) to LLVM syntax, then will be easier to adjust it

**Comment by stuartarchibald** [comment]
> Thanks for testing, 

No problem.

> would be good to have most complicated function to test lexer on.

This one is pretty involved, it's designed for testing our IR-diff tool: https://github.com/numba/numba/blob/868b8e3e8d034dac0440b75ca31595e07f632d27/numba/tests/test_ir.py#L349-L385

I've extracted it here:
```python
from numba.core import compiler
import numpy as np

_GLOBAL = 1234
def gen():
    _FREEVAR = 0xCAFE

    def foo(a, b, c=12, d=1j, e=None):
        f = a + b
        a += _FREEVAR
        g = np.zeros(c, dtype=np.complex64)
        h = f + g
        i = 1j / d
        if np.abs(i) > 0:
            k = h / i
            l = np.arange(1, c + 1)
            with objmode():
                print(e, k)
            m = np.sqrt(l - g)
            if np.abs(m[0]) < 1:
                n = 0
                for o in range(a):
                    n += 0
                    if np.abs(n) < 3:
                        break
                n += m[2]
            p = g / l
            q = []
            for r in range(len(p)):
                q.append(p[r])
                if r > 4 + 1:
                    with objmode(s='intp', t='complex128'):
                        s = 123
                        t = 5
                    if s > 122:
                        t += s
                t += q[0] + _GLOBAL

        return f + o + r + t + r + a + n

    return foo

ir = compiler.run_frontend(gen())
print(ir.dump())
```
and it looks good to me. This one runs a load of the more recently supported features, including a lot of transformation/code generation:
```python
from numba import config
config.DEBUG_PRINT_AFTER='ir_legalization'
from numba import literal_unroll, njit, gdb
import numpy as np

import operator
from functools import reduce
from numba.typed import List

@njit(['int64()',])
def newish_features():

    def mul_n(x, multiplier):
        return x * multiplier

    V = 20
    def greater_than_V(x):
        return x > V

    reduce_lambda = lambda x, y: (x * 2) + y

    gdb()

    a = [x ** 2 for x in range(10)]
    n = len(a)
    try:
        return a[3]
    except Exception:
        print("caught exception")
        return -1

    tup = (1, 'a', 2j)
    for i in literal_unroll(tup):
        print(i)

    return reduce(reduce_lambda, filter(greater_than_V, map(mul_n, a, range(n))))
```
and again seems reasonable.

> If you will explain your changes(or put a link to the source) to LLVM syntax, then will be easier to adjust it

Here's an example:
```python
from numba import config
config.DUMP_OPTIMIZED=1
from numba import njit
import numpy as np

@njit
def foo(x):
    y = np.sin(x)
    return x + y[0]

foo(np.ones(5))
```
produces something like:
highlighting_issue2

the LLVM Lexer in use is the one here: https://github.com/numba/numba/blob/868b8e3e8d034dac0440b75ca31595e07f632d27/numba/core/lowering.py#L193

LLVM language reference is here http://llvm.org/docs/LangRef.html, I wonder if part of what is observed is possibly due to the Lexer not having a specific LLVM IR version, and this being a mis-match.

**Comment by 1844144** [comment]
Looks like you are using Gas lexer for optimized dump (which for Gas (AT&T) assembly code) is it ok?
https://github.com/numba/numba/blob/868b8e3e8d034dac0440b75ca31595e07f632d27/numba/core/codegen.py#L34-L36

**Comment by stuartarchibald** [comment]
> Looks like you are using Gas lexer for optimized dump (which for Gas (AT&T) assembly code) is it ok?
> 
> https://github.com/numba/numba/blob/868b8e3e8d034dac0440b75ca31595e07f632d27/numba/core/codegen.py#L34-L36
> 
> Changing to LlvmLexer gives desired result. Is there a cases when you need Gas lexer?

Oops, that'd explain it, needs two code paths! RE GAS Lexer, yes, we often want the LLVM IR and the assembly next to each other, setting the env var `NUMBA_DEBUG=1` is common when debugging and it dumps the Numba IR, LLVM IR and assembly.

**Comment by gmarkall** [comment]
>  @gmarkall wonder if a PTX (and maybe later SASS) lexer would be useful for CUDA dumps?

It would be - I find that GNU Assembler syntax already works reasonably well for PTX in vim, so probably the `GasLexer` would be a good starting point: https://pygments.org/docs/lexers/#pygments.lexers.asm.GasLexer

**Pull request merged**